### PR TITLE
Autocompletion for qmk.

### PR DESCRIPTION
**Pull request by undg**
- [*] This compdef is not already available in zsh.
- [*] This compdef is not already available in their original project.
- [*] I am the original author, or I have authorization to submit this work.
- [*] This is a finished work.
- [*] It has a header containing authors, status and origin of the script.
- [*] It has a license header or I accept that it will be licensed under the terms of the Zsh license.

Completion for all arguments and for many options in [QMK CLI](https://beta.docs.qmk.fm/cli/cli)
image
image

### CODE REVIEW

File: src/_qmk [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    _arguments \
+        '-h[Print help text.]' \
+        '--help[Print help text.]' \
+        "-n[Don't actually build, just show the make command to be run.]::_qmk_compile" \
+        "--dry-run[Don't actually build, just show the make command to be run.]::_qmk_compile" \
+        '-km[The keymap to build a firmware for. Ignored when a configurator export is supplied.]::_qmk_compile' \
+        '--keymap[The keymap to build a firmware for. Ignored when a configurator export is supplied.]::_qmk_compile' \
+        '-kb[The keyboard to build a firmware for. Ignored when a configurator export is supplied.]::_qmk_compile' \
+        '--keyboard[The keyboard to build a firmware for. Ignored when a configurator export is supplied.]::_qmk_compile' \
+        '*::configuratorExport:{_values "json files" *.json}'
```
Comment (syohex) [inline src/_qmk:53]:
```suggestion
        '*::configuratorExport:_files -g "*.json"'
```

I got following error when I complete at `qmk compile`.

```
% qmk compile # tab press
_values:compvalues:11: not enough arguments
```

File: src/_qmk [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        "--dry-run[Don't actually build, just show the make command to be run.]::_qmk_flash" \
+        '-km[The keymap to build a firmware for. Ignored when a configurator export is supplied.]::_qmk_flash' \
+        '--keymap[The keymap to build a firmware for. Ignored when a configurator export is supplied.]::_qmk_flash' \
+        '-kb[The keyboard to build a firmware for. Ignored when a configurator export is supplied.]::_qmk_flash' \
+        '--keyboard[The keyboard to build a firmware for. Ignored when a configurator export is supplied.]::_qmk_flash' \
+        "-bl[The flash command, corresponding to qmk's make options of bootloaders.]::_qmk_flash"\
+        "--bootloader[The flash command, corresponding to qmk's make options of bootloaders.]::_qmk_flash"\
+        '-b[List the available bootloaders.]::_qmk_flash' \
+        '--bootloaders[List the available bootloaders.]::_qmk_flash' \
+        '*::configuratorExport:{_values "json files" *.json}'
```
Comment (syohex) [inline src/_qmk:76]: ditto

File: src/_qmk [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+_qmk_json2c(){
+    _arguments \
+        '-q[Quiet mode, only output error messages]'\
+        '--quiet[Quiet mode, only output error messages]'\
+        '-o[File to write to]'\
+        '--output[File to write to]'\
+        '-h[Print help text.]'\
+        '--help[Print help text.]'\
+        '*::configuratorExport:{_values "json files" *.json}'
```
Comment (syohex) [inline src/_qmk:105]: ditto

**Comment by undg** [comment]
@syohex Thank's for your suggestion. 

I've checked all commands against the newest version QMK CLI and it looks like now we have all commands and flags.

**Comment by undg** [comment]
...And renamed commits.

**Review by syohex — APPROVED** [review]

File: src/_qmk [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    _arguments \
+        "-h[Print help text.]" \
+        "--help[Print help text.]" \
+}
+
+_qmk_chibios-confmigrate(){
+    _arguments \
+        "-h[Print help text.]" \
+        "--help[Print help text.]" \
+        "-f: [Re-migrates an already migrated file, even if it doesn't detect a full ChibiOS config.]" \
```
Comment (syohex) [inline src/_qmk:175]:
```suggestion
        "-f[Re-migrates an already migrated file, even if it doesn't detect a full ChibiOS config.]" \
```

Is `: ` after option necessary ? I suppose if there is it, completion document does not show.

File: src/_qmk [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+#  * precondition (https://github.com/precondition)
+#  * undg (https://github.com/undg)
+#
+# ------------------------------------------------------------------------------
+
+
+_qmk_compile(){
+    _arguments \
+        "-h[Print help text.]" \
+        "--help[Print help text.]" \
```
Comment (syohex) [inline src/_qmk:47]:
```suggestion
        {-h,--help}"[Print help text.]" \
```

Short option and long option can be combined.

**Review by syohex — COMMENTED** [review]
Sorry I mistook to approve. But I have some comments.

File: src/_qmk (hunk 5 above)
Comment (undg) [reply src/_qmk:47]: I was woned if there is shorthand for that, thx.

File: src/_qmk (hunk 4 above)
Comment (undg) [reply src/_qmk:175]: No, it's a typo.

File: src/_qmk [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        {-h,--help}"[Print help text.]" \
+        {-n,--no}"[Answer no to all questions.]" \
+        {-y,--yes}"[Answer yes to all questions.]" \
+}
+
+_qmk_json2c(){
+    _arguments \
+        {-h,--help}"[Print help text.]" \
+        {-q,--quiet}"[Quiet mode, only output error messages]" \
+        {-o,--output}"[File to write to]" \
```
Comment (syohex) [inline src/_qmk:84]:
```suggestion
        {-o,--output}"[File to write to]:output_file:_files" \
```

I suppose that options which takes file name should complete path name. How about ?

File: src/_qmk [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        {-b,--branch}"[The branch to clone]" \
+        {-y,--yes}"[Answer yes to all questions.]" \
+        {-n,--no}"[Answer no to all questions.]" \
+        "--baseurl[The URL all git operations start from]" \
+}
+
+_qmk_info(){
+    _arguments \
+        {-h,--help}"[Print help text.]" \
+        {-f,--format}"[Format to display the data in (friendly, text, json)(Default: friendly).]" \
```
Comment (syohex) [inline src/_qmk:119]:
```suggestion
        {-f,--format}"[Format to display the data in (friendly, text, json)(Default: friendly).]: :(friendly text json)" \
```

File: src/_qmk [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+_qmk_command(){
+    local -a _qmk_cmds
+    _qmk_cmds=(
+        "config: Read and write configuration settings." \
+        "clone: Clone a qmk_firmware fork." \
+        "env: Prints environment information." \
+        "setup: Setup your computer for qmk_firmware." \
+        "c2json: Creates a keymap.json from a keymap.c file." \
```
Comment (syohex) [inline src/_qmk:172]: c2json subcommand has some command line flags but `_qmk_c2json` is not defined. Is this right ?

https://beta.docs.qmk.fm/cli/cli_commands

Comment (undg) [reply src/_qmk:188]: I've added c2json and clone. They both were missing.

**Review by syohex — APPROVED** [review]
@undg thanks looks good. Cloud you squash commits before merging ?

**Comment by undg** [comment]
sure I can, thx for help.

**Pull request merged**